FIN-25: Implement backend for existing screens

FedericoLeiva12 opened the pull request:
## 📝 Description
- Implemented `accounts`, `transactions`, `categories` and `users` routes, schemas and services in backend.
- Adjusted first login flow to correctly create account in database.
- Generated schemas with all new models.
- Created seed script to generate initial database data.
- Added backend docs.

## TODO (pending)
Refactor home screen to be able to manage new backend types and replace mock data with real backend data.

- [ ] Bug fix
- [x] New feature
- [x] Refactoring
- [x] Documentation update
- [ ] Other (please specify)

## ✅ Checklist
- [x] Code has been tested and works as expected (White/ Darkmode)
- [x] Documented Code (Please read your code, think about what might be unclear, and explain your code.)
- [x] Clear Pull Request title and description (Explain "What have you changed?")
- [x] All new text uses translations. If you added any UI text to the Finance.io frontend, ensure it's internationalized
- [x] TypeScript is used with clear, explicit types
- [x] NativeWind/Tailwind CSS is used consistently
- [x] Only approved icon packs and libraries are used
- [x] Relevant tests have been added or updated
- [x] Documentation has been updated (README, Wiki, etc.)
- [x] Linter/Formatter runs without errors
- [x] Branch is up-to-date with `master`/`develop`

## 🔍 Screenshots / GIFs (optional)

## 🔗 Related Issues
Closes #65 
Closes #68 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Account type classification system with Private, Business, Savings, and Shared account categories.
  * Transaction tracking and management with category and currency support.
  * Multi-currency account support with automatic currency handling.
  * Enhanced account dashboard with organized account sections by type.

* **Documentation**
  * Backend development guidelines documenting tech stack, authentication, error handling, and feature-adding checklist.

* **Improvements**
  * Improved error handling and standardized error messaging across the application.
  * Updated localization across 20+ languages to support new account and transaction features.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

FedericoLeiva12 committed:
recreated backend database

FedericoLeiva12 committed:
feat: add seed script to package.json for database seeding

FedericoLeiva12 committed:
feat: implement account types and enhance accounts management

- Added account types to the database schema and created related migrations.
- Introduced new API routes for managing accounts and retrieving user data.
- Updated the frontend to handle account creation and retrieval with type support.
- Added error handling for account-related operations.
- Created a backend development guide for consistency and best practices.

FedericoLeiva12 committed:
docs: update backend development guide and improve code formatting

- Enhanced the backend development documentation with additional guidelines and best practices.
- Improved code formatting for better readability in various files, including schema and router definitions.
- Added missing newlines in JSON files for consistency.
- Updated mock data structure to align with new account properties.

FedericoLeiva12 committed:
Merge branch 'master' of https://github.com/The-Creative-Programming-Group/finance-io into federicoleiva/fin-25-implement-backend-for-existing-screens

FedericoLeiva12 commented on src/app/(auth)/sign-in.tsx:
```diff
@@ -99,7 +99,7 @@ export default function Page() {
       });
       if (signInAttempt.status === "complete") {
         await setActive({ session: signInAttempt.createdSessionId });
-        router.replace("/start");
```
For reviewers:
`/` screen is currently managing redirect for logged users.

FleetAdmiralJakob replied:
Please comment the hell out of this behaviour. Right now it's not really clear what is happening in the app

FedericoLeiva12 committed:
fix: update mock data structure for consistency

- Changed the property name from 'bankName' to 'name' in the mock dashboard data to align with the updated data structure.

FedericoLeiva12 committed:
refactor: improve query conditions and enhance user handling

- Refactored the query conditions in the transactions router for better readability and maintainability by using an array to build conditions.
- Updated the user retrieval logic in the users router to use `onConflictDoNothing` for inserting users, ensuring no duplicates are created.
- Enhanced the backend development documentation with clearer examples and improved code formatting.

Gamius00 commented on src/app/index.tsx:
```diff
@@ -140,10 +141,28 @@ export default function Index() {
+    return <Redirect href={"./(tabs)/banking"} />;
   }
 
+  /// TODO: Add loading state
```
Please remove it and create a Ticket for it

FleetAdmiralJakob replied:
In Linear ofc

FleetAdmiralJakob replied:
https://linear.app/financeio/issue/FIN-31/add-loading-state-to-the-landing-page

Gamius00 commented on src/data/mockData.ts:
```diff
@@ -5,10 +5,11 @@ import {
 import { CircleArrowRightIcon, LandmarkIcon } from "lucide-react-native";
 
-export const mockDashboardData: DashboardData = {
+// TODO: Remove mock data
```
Please remove the mock Data

FleetAdmiralJakob replied:
From now on we only use real data

Gamius00 commented on scripts/seed.ts:
```diff
@@ -0,0 +1,101 @@
+      `Inserted currencies: ${toInsert.map((c) => c.code).join(", ")}`,
+    );
+  } else {
+    console.log("Currencies already seeded");
```
Please remove the console log

Gamius00 commented on scripts/seed.ts:
```diff
@@ -0,0 +1,101 @@
+
+  if (toInsert.length > 0) {
+    await db.insert(currenciesTable).values(toInsert);
+    console.log(
```
Remove the console.log

FedericoLeiva12 replied:
console.logs in seed.ts are intended to be there. This script is the one that fills the database with default information, and the logs helps the developer to understand what it did.

FleetAdmiralJakob replied:
Yeah, I agree with Federico. 🤣

Gamius00 commented on scripts/seed.ts:
```diff
@@ -0,0 +1,101 @@
+
+  const codes = currencies.map((c) => c.code);
+
+  const existing = await db
```
Please comment it

Gamius00 commented on scripts/seed.ts:
Please make more comments, for every function ... etc.

FleetAdmiralJakob replied:
yes

Gamius00 commented on src/app/index.tsx:
```diff
@@ -140,10 +141,28 @@ export default function Index() {
 
-  if (isSignedIn) {
-    return <Redirect href={"../home"} />;
+  const { data: accounts, isLoading: isLoadingAccounts } =
```
Comment it

Gamius00 reviewed (commented):
Please comment: 
- Functions 
- Variables and States 

But you dont have to comment your schema or something.

Gamius00 commented on docs/backend-development.md:
Please remove this read me and add it to our wiki: https://github.com/[The-Creative-Programming-Group/finance-io](https://github.com/The-Creative-Programming-Group/finance-io/wiki)/wiki

FleetAdmiralJakob replied:
We want to have everything in one place, and we figured out that the wiki is the best for that.

FleetAdmiralJakob replied:
A problem we haven't thought about yet then is that if we put everything in the wiki AIs won't be able to access all of that information......... hmmmm......

Gamius00 replied:
You mean its important for like GH Copilot if it creates a PR?

FleetAdmiralJakob replied:
Not only that but also if it works in your IDE. Then it has context and knows what to do.

FleetAdmiralJakob commented on scripts/seed.ts:
```diff
@@ -0,0 +1,101 @@
+
+async function seedCurrencies() {
+  const currencies = [
+    { code: "USD", name: "US Dollar" },
+    { code: "EUR", name: "Euro" },
+    { code: "ARS", name: "Argentine Peso" },
```
Maybe we create a type here with all the currencies we want to support and also use that in the schema. Unite everything in one file so it's easy to change and then reuse it everywhere

FleetAdmiralJakob replied:
Yes, let's do it with an enum

FleetAdmiralJakob commented on src/db/schema.ts:
```diff
@@ -1,17 +1,110 @@
 });
+
+// Currencies
+export const currenciesTable = pgTable(
+  "currencies",
+  {
+    id: uuid().defaultRandom().primaryKey(),
+    name: varchar({ length: 100 }).notNull(),
+    code: varchar({ length: 10 }).notNull(), // e.g. USD, EUR, ARS
+  },
+  (table) => ({
+    codeUniqueIdx: uniqueIndex("currencies_code_unique_idx").on(table.code),
+  }),
+);
```
Why do we need a table storing different currencies?

FleetAdmiralJakob commented on src/db/schema.ts:
```diff
@@ -1,17 +1,110 @@
+    slugUniqueIdx: uniqueIndex("account_types_slug_unique_idx").on(table.slug),
+  }),
+);
+
```
Can you name an example of an account type? Why can't this be an enum or something? Why should this be a table?

FleetAdmiralJakob commented on src/db/schema.ts:
```diff
@@ -1,17 +1,110 @@
+  }),
+);
+
+// Categories
+export const categoriesTable = pgTable(
+  "categories",
+  {
+    id: uuid().defaultRandom().primaryKey(),
+    name: varchar({ length: 100 }).notNull(),
+    slug: varchar({ length: 120 }).notNull(),
+  },
+  (table) => ({
+    slugUniqueIdx: uniqueIndex("categories_slug_unique_idx").on(table.slug),
+  }),
+);
```
Same thing with currencies and account types: Does this need to be its own table?

FleetAdmiralJakob commented on scripts/seed.ts:
```diff
@@ -0,0 +1,101 @@
+}
+
+async function seedCategories() {
+  const categories = [
+    { name: "Groceries", slug: "groceries" },
+    { name: "Restaurants", slug: "restaurants" },
+    { name: "Transport", slug: "transport" },
+    { name: "Bills", slug: "bills" },
+    { name: "Entertainment", slug: "entertainment" },
+  ];
```
Same question as with the currencies

FleetAdmiralJakob commented on src/server/routers/users.ts:
We need to make sure what happens if Clerk and our DB get out of sync.

Why did we choose not to use Clerk Webhooks to sync both?
https://clerk.com/docs/guides/development/webhooks/overview

FleetAdmiralJakob commented on src/types/index.ts:
```diff
@@ -1,11 +1,62 @@
+ * Our database only saves the Clerk user id, so we need to get the user details from Clerk.
+ * When we request the user defails from our backend it automatically gets the user details from Clerk.
+ */
+export interface ClerkUser {
+  id: string;
+  firstName: string;
+  lastName: string;
+  emails: string[];
+  phoneNumbers: string[];
+}
```
This needs to be kept in sync if we change settings in the Clerk dashboard. Also, we don't have phone number support AFAIK.

FleetAdmiralJakob commented on src/types/index.ts:
```diff
@@ -1,11 +1,62 @@
+export interface TransactionWithCategoryAndAccount extends Transaction {
+  category: Category;
+  account: Account;
 }
```
Can't these types be inferred or something from Drizzle e.g.

Like so we don't violate DRY.

FleetAdmiralJakob reviewed (changes requested):
Could you please create a video and upload it here showing how the app looks and works with these features added?

FleetAdmiralJakob committed:
format correctly

FleetAdmiralJakob committed:
Merge branch 'master' into federicoleiva/fin-25-implement-backend-for-existing-screens

FleetAdmiralJakob committed:
comment the behaviour

FleetAdmiralJakob committed:
improve the names of the commands in the package.json

FleetAdmiralJakob committed:
fix issues with the migrations

FleetAdmiralJakob committed:
fix typos

FleetAdmiralJakob committed:
fix typos

FleetAdmiralJakob committed:
remove unnecessary optional chaining

FleetAdmiralJakob committed:
improve commenting

FleetAdmiralJakob committed:
fix error handling and unused code

FleetAdmiralJakob commented on src/server/api/trpc.ts:
```diff
@@ -123,14 +127,24 @@ export const publicProcedure = t.procedure;
+    throw errors.notAuthenticated();
   }
+
+  let user;
+  try {
+    user = await clerkClient.users.getUser(ctx.userId);
+  } catch (err) {
+    console.error("Error fetching user from Clerk:", err);
+    throw errors.notAuthenticated();
+  }
```
@coderabbitai Does it make sense to use lru-cache in a serverless environment?

FleetAdmiralJakob replied:
https://github.com/coderabbitai We will use a serverless environment with a 99% probability. Please reevaluate your answer and tips. And update your review comment.

FleetAdmiralJakob replied:
@coderabbitai We will use a serverless environment with a 99% probability. Please reevaluate your answer and tips. And update your review comment.

FleetAdmiralJakob committed:
fix the sign-up and account creation process

- improve types
- removed "account types" table

FleetAdmiralJakob committed:
switched to using the real data on the dashboard page

- removed mockData

FleetAdmiralJakob commented:
https://github.com/user-attachments/assets/7e80a7b1-b0f4-42b7-a0b4-348c8ee0528c

FleetAdmiralJakob commented on src/components/AccountItem.tsx:
```diff
@@ -27,12 +27,15 @@ export const AccountItem: React.FC<AccountItemProps> = ({
       })
     : icon;
 
+   */
```
Nah, we will need this code later. But this is not the right PR to implement this kind of thing. That's why I commented it out to make it easier in the future to add it back in again

FleetAdmiralJakob commented on src/components/AccountItem.tsx:
```diff
@@ -5,7 +5,7 @@ import AppText from "./ui/AppText";
 
 interface AccountItemProps {
-  icon: React.ReactNode;
+  // icon: React.ReactNode;
```
Nah, we will need this code later. But this is not the right PR to implement this kind of thing. That's why I commented it out to make it easier in the future to add it back in again

FleetAdmiralJakob committed:
make intent clear

FleetAdmiralJakob committed:
Improve type safety and efficiency in account filtering

- cleaned up the types

FleetAdmiralJakob committed:
added some todos

FleetAdmiralJakob committed:
Add newline at end of file.

The file is missing a trailing newline, which is required by POSIX standards and may cause issues with some tools.

FleetAdmiralJakob committed:
resolved comments

cleaned up the scripts structure
